api/filters/reference_instance_filter.go

```diff
@@ -0,0 +1,214 @@
+		return next.Handle(req)
+	}
+	instance := dbInstanceObject.(*types.ServiceInstance)
+
+	isReferencePlan, err := f.isReferencePlan(ctx, instance.ServicePlanID)
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	_, err = f.isValidPatchRequest(req, instance)
+	if err != nil {
+		return nil, err
+	}
+	return next.Handle(req)
+}
+
+func (f *referenceInstanceFilter) isReferencedShared(ctx context.Context, referencedInstanceID string) (bool, error) {
+	byID := query.ByField(query.EqualsOperator, "id", referencedInstanceID)
+	dbReferencedObject, err := f.repository.Get(ctx, types.ServiceInstanceType, byID)
+	if err != nil {
+		return false, util.HandleStorageError(err, types.ServiceInstanceType.String())
+	}
+	referencedInstance := dbReferencedObject.(*types.ServiceInstance)
+
+	if !*referencedInstance.Shared {
+		return false, util.HandleInstanceSharingError(util.ErrReferencedInstanceNotShared, referencedInstanceID)
+	}
+	return true, nil
+}
+
+func (f *referenceInstanceFilter) isValidPatchRequest(req *web.Request, instance *types.ServiceInstance) (bool, error) {
+	newPlanID := gjson.GetBytes(req.Body, planIDProperty).String()
+	if instance.ServicePlanID != newPlanID {
+		return false, util.HandleInstanceSharingError(util.ErrChangingPlanOfReferenceInstance, instance.Name)
+	}
+
+	parametersRaw := gjson.GetBytes(req.Body, "parameters").Raw
+	if parametersRaw != "" {
+		return false, util.HandleInstanceSharingError(util.ErrChangingParametersOfReferenceInstance, instance.Name)
+	}
+
+	return true, nil
+}
+
+func (f *referenceInstanceFilter) isReferencePlan(ctx context.Context, servicePlanID string) (bool, error) {
+	dbPlanObject, err := f.getObjectByID(ctx, types.ServicePlanType, servicePlanID)
+	if err != nil {
+		return false, err
+	}
+	plan := dbPlanObject.(*types.ServicePlan)
+	return plan.Name == constant.ReferencePlanName, nil
+}
+
+func (f *referenceInstanceFilter) getObjectByID(ctx context.Context, objectType types.ObjectType, resourceID string) (types.Object, error) {
+	byID := query.ByField(query.EqualsOperator, "id", resourceID)
+	dbObject, err := f.repository.Get(ctx, objectType, byID)
+	if err != nil {
+		return nil, util.HandleStorageError(err, objectType.String())
+	}
+	return dbObject, nil
+}
+
+func (f *referenceInstanceFilter) validateOwnership(req *web.Request) error {
+	ctx := req.Context()
+	callerTenantID := gjson.GetBytes(req.Body, "context."+f.tenantIdentifier).String()
+	path := fmt.Sprintf("parameters.%s", constant.ReferencedInstanceIDKey)
+	referencedInstanceID := gjson.GetBytes(req.Body, path).String()
+	byID := query.ByField(query.EqualsOperator, "id", referencedInstanceID)
+	dbReferencedObject, err := f.repository.Get(ctx, types.ServiceInstanceType, byID)
+	if err != nil {
+		return util.HandleStorageError(err, types.ServiceInstanceType.String())
+	}
+	instance := dbReferencedObject.(*types.ServiceInstance)
+	referencedOwnerTenantID := instance.Labels[f.tenantIdentifier][0]
```

Comment:
what will happen if instance.Labels be without f.tenantIdentifier?  by default sm is not multi tenant.
Do we have a test for this case?